### ISSUE TITLE
Do not check $cse variables for event iteration

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9598]
We are currently trying to solve a problem on an application in thermo-fluid systems that eventually fails because of a bogus terminal event iteration. The details of that are inessential, so they are not reported here. The bottom line is that the simulation fails because of a bogus non-converging terminal event iteration that is triggered by some $cse discrete variables that keep changing at each iteration.

The solution procedure sketched in [Appendix B](https://specification.modelica.org/maint/3.5/modelica-dae-representation.html) of the Modelica Specification requires to check if `pre(m) == m` at the end of the event iteration, to decide whether to iterate again or not. Now, `m` is defined as "Modelica variables of discrete-valued types (Boolean, Integer, etc) which are unknown.". 

Now, $cse variables are additional entities introduced by OMC for the sake of efficient computation of function calls, but they are not part of the Modelica model. They just contain outputs of function calls that may only be used as inputs to other function calls, without appearing explicitly as variables in the Modelica model. 

Hence, **any discrete $cse variable should not belong to the set of discrete variables that is checked for the even iteration termination condition `pre(m) == m`.**

This is consistent with the well-know principle that that function calls per se should not generate events.

You can use this (somewhat artificial) MWE to check that OMC fulfills this specification:

```modelica
model TestBogusEventIteration
  package Medium = Modelica.Media.Water.StandardWater;
  Medium.AbsolutePressure p(start = 1.01e5, fixed = true);
  Medium.AbsolutePressure p0 = 1e5;
  Medium.Density rho;
  Medium.Temperature T;
  Medium.SpecificEnthalpy h, hv;  
  Medium.MassFlowRate w;
equation
  der(p) = 1e3;
  p - p0 = w^2/rho;
  hv = Medium.dewEnthalpy(Medium.setSat_p(p));
  T = Medium.temperature_ph(p,h);
  rho = Medium.density_ph(p,h);
  if terminal() then 
    h = hv + 1e-6;
  else 
    h = hv;
  end if; 
annotation(
    __OpenModelica_simulationFlags(lv = "LOG_EVENTS_V"));
end TestBogusEventIteration;
```
During the terminal iteration, "check for discrete changes at time=1" should find no discrete change, and no event iteration should be performed.

**Comment by casella** [comment]
@AnHeuermann, @phannebohm if this is a quick fix it would be nice to have it soon, otherwise let's discuss what is needed to get that.

**Comment by casella** [comment]
Keeping @AndreaBartolini and @pepmts in the loop.

**Comment by AnHeuermann** [comment]
We are not differentiating between CSE and non-CSE variables inside the C runtime.
So in function `checkForDiscreteChanges` there is no way to tell if `data->localData[0]->integerVars[i]` is a CSE variable or not.

To solve this one would need to save the information which variables are CSV variables and use it in `checkForDiscreteChanges`.
But I don't see an quick way to do that.

**Comment by casella** [comment]
@AnHeuermann I'm not into the details of the data structure, but looking at runtime error logs (e.g. `LOG_NLS_V`) it is clear to me that the runtime has access to the variable _names_. You can just tell whether or not a discrete variable is a CSE by checking if its name starts with `$cse`.

Do I miss something?

**Comment by AnHeuermann** [comment]
Yes, but parsing names every time an integrator step is done would be extremely slow. So we want to do it only once.
But it is still not quick to parse the list of variable names and search for a string in it. Generating an array of CSV variable indices during Code generation should be the most performant way, but I'm not sure if we have a list of CSV variables during SimCode with the matching runtime indices.

**Comment by casella** [comment]
> Yes, but parsing names every time an integrator step is done would be extremely slow

I don't think so. There is no real parsing involved, just calling `strcmp(data->localdata[0]->name[i], "$cse", 4)` or something like that. `strcmp()` is very, very efficient, this translates into something as 
```C
for(i = 0; i < len; i++)
  if(*char1++ != *char2++)
    return 0;
return 1;
```
that eventually is turned into a handful of assembly instruction calls. 

Also, you don't need to do that at every integrator time step, only when doing event iteration. Basically, at the end of the event iteration code you already have something like this (in pseudo-code)
```
iterate = 0;
for i in <list of discrete variable indeces> loop
  if var[i] != pre_var[i] then 
    iterate = 1;
    break;
  end if;
end for;
if not iterate then return else <do another event iteration>;
```
You just need to turn it into
```
iterate = 0;
for i in <list of discrete variable indeces> loop
  if var[i] != pre_var[i] and not strcmp(data->localdata[0]->name, "$cse", 4) then 
    iterate = 1;
    break;
  end if;
end for;
if not iterate then return else <do another event iteration>;
```
No big deal. Do I miss something?

**Comment by AnHeuermann** [comment]
I guess you are right.
Looping over `data->modelData->realVarsData->info->name` should do the trick. (Replace `realVarsData` with the other types).

I or @phannebohm will add something for it to the runtime.

**Comment by casella** [comment]
Great, thanks!

**Comment by casella** [comment]
Ping 😅

**Comment by casella** [comment]
Thanks @phannebohm!
